### PR TITLE
Implement fallback for consistent reads from cache

### PR DESCRIPTION
**Pull request by serathius**
/kind bug

```release-note
NONE
```
cc @liggitt

### CODE REVIEW

**Comment by serathius** [comment]
cc @wojtek-t @deads2k

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -848,7 +848,8 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 		preparedKey += "/"
 	}
 	requestWatchProgressSupported := etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
-	if resourceVersion == "" && utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && requestWatchProgressSupported {
+	consistentRead := resourceVersion == "" && utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && requestWatchProgressSupported
+	if consistentRead {
 		listRV, err = storage.GetCurrentResourceVersionFromStorage(ctx, c.storage, c.newListFunc, c.resourcePrefix, c.objectType.String())
 		if err != nil {
 			return err
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:855]: Don't we have to fallback here?

Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:855]: No, this is the request we sent to storage to get consistent revision from etcd. Only when we list from watch cache and get a timeout error we fallback.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 2]
```diff
@@ -888,8 +889,15 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 
 	objs, readResourceVersion, indexUsed, err := c.listItems(ctx, listRV, preparedKey, pred, recursive)
 	if err != nil {
+		if consistentRead && storage.IsTooLargeResourceVersion(err) {
+			metrics.RecordConsistentReadSuccess(false)
+			return c.storage.GetList(ctx, key, opts, listObj)
+		}
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:895]: A test would probably be a good idea.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:855]: okie dokie.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 2 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:895]: Working on it

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go [hunk 3]
```diff
@@ -167,6 +167,15 @@ var (
 			StabilityLevel: compbasemetrics.ALPHA,
 			Buckets:        []float64{0.005, 0.025, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3},
 		}, []string{"resource"})
+
+	consistentReadSuccess = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "consistent_read_success",
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:175]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -888,8 +889,15 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 
 	objs, readResourceVersion, indexUsed, err := c.listItems(ctx, listRV, preparedKey, pred, recursive)
 	if err != nil {
+		if consistentRead && storage.IsTooLargeResourceVersion(err) {
+			metrics.RecordConsistentReadSuccess(false)
+			return c.storage.GetList(ctx, key, opts, listObj)
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: Done

**Comment by liggitt** [comment]
lgtm

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 5]
```diff
@@ -888,8 +889,15 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 
 	objs, readResourceVersion, indexUsed, err := c.listItems(ctx, listRV, preparedKey, pred, recursive)
 	if err != nil {
+		if consistentRead && storage.IsTooLargeResourceVersion(err) {
+			metrics.RecordConsistentRead(c.resourcePrefix, false)
+			return c.storage.GetList(ctx, key, opts, listObj)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:894]: How can we know what percentage of these calls succeeded?

Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:894]: `apiserver_watch_cache_consistent_read_total{success="true"}/(apiserver_watch_cache_consistent_read_total{success="true"}+apiserver_watch_cache_consistent_read_total{success="false"})

Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:894]: Please see https://github.com/kubernetes/community/pull/7708/files

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:894]: > apiserver_watch_cache_consistent_read_total{success="true"}/(apiserver_watch_cache_consistent_read_total{success="true"}+apiserver_watch_cache_consistent_read_total{success="false"})

Intuitively, this appears to be the percent of consistent reads that result calling line 894.  But I don't see how this discriminates between this list succeeding and failing.  Perhaps walking through examples will help.


1. consistent read fails on something other than istoolarge: I think this is not included in `ConsistentReadTotal` at all.
2. consistent read success: I think this is `ConsistentReadTotal` labeled with `success=true`
3. consistent read fail on istoolarge, backup list succeeds: I think this is `ConsistentReadTotal` labeled with `success=false`
4. consistent read fail on istoolarge, backup list fails: I think this is `ConsistentReadTotal` labeled with `success=false`

Answers I'd like to be able to get from metrics:
1. there have been four total consistent read requests.  I think currently I will think there are three, not four.
2. 25% of consistent reads succeed.  I think currently I will believe that is 33%.
3. 50% of backup lists have succeeded.  I think currently there is no way to determine.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 6]
```diff
@@ -888,8 +889,19 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 
 	objs, readResourceVersion, indexUsed, err := c.listItems(ctx, listRV, preparedKey, pred, recursive)
 	if err != nil {
+		if consistentRead {
+			if storage.IsTooLargeResourceVersion(err) {
+				metrics.ConsistentReadTotal.WithLabelValues(c.resourcePrefix, "fallback").Add(1)
+				return c.storage.GetList(ctx, key, opts, listObj)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:895]: Reworking my previous example on this updated code.  We have these requests

1.  consistent read fails on something other than istoolarge: I think this `ConsistentReadTotal` labelled with `failure`.
2. consistent read success: I think this is `ConsistentReadTotal` labeled with `success`
3. consistent read fail on istoolarge, backup list succeeds: I think this is `ConsistentReadTotal` labeled with `fallback`
4. consistent read fail on istoolarge, backup list fails: I think this is `ConsistentReadTotal` labeled with `fallback`

Answers I'd like to be able to get from metrics:

1. there have been four total consistent read requests. Correct.
2. 25% of consistent reads succeed. Correct.
3. 50% of backup lists have succeeded. I think currently there is no way to determine.

**Comment by enj** [comment]
I ran `make test-integration WHAT=./test/integration/storageversionmigrator KUBE_TEST_ARGS="-v -run CRD"` on 319306218eaf275ad89c25539de0386aac616e9e with the following diff:

```patch
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
index e3eecc322d9..d71839d3e98 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
@@ -892,6 +892,8 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 		if consistentRead {
 			if storage.IsTooLargeResourceVersion(err) {
 				metrics.ConsistentReadTotal.WithLabelValues(c.resourcePrefix, "fallback").Add(1)
+				m, _ := metrics.LocalRegistry.Gather()
+				klog.Errorf(">>>>>>>>>> Saw fallback for %s\n%s", c.resourcePrefix, m)
 				return c.storage.GetList(ctx, key, opts, listObj)
 			} else {
 				metrics.ConsistentReadTotal.WithLabelValues(c.resourcePrefix, "failure").Add(1)
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go b/staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go
index 90f4b54fbe8..614148e9680 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go
@@ -180,6 +180,8 @@ var (
 
 var registerMetrics sync.Once
 
+var LocalRegistry = compbasemetrics.NewKubeRegistry()
+
 // Register all metrics.
 func Register() {
 	// Register the metrics.
@@ -198,6 +200,7 @@ func Register() {
 		legacyregistry.MustRegister(WatchCacheInitializations)
 		legacyregistry.MustRegister(WatchCacheReadWait)
 		legacyregistry.MustRegister(ConsistentReadTotal)
+		LocalRegistry.MustRegister(ConsistentReadTotal)
 	})
 }
 
diff --git a/test/images/agnhost/crd-conversion-webhook/converter/framework.go b/test/images/agnhost/crd-conversion-webhook/converter/framework.go
index c16a684f0d3..707999c86d7 100644
--- a/test/images/agnhost/crd-conversion-webhook/converter/framework.go
+++ b/test/images/agnhost/crd-conversion-webhook/converter/framework.go
@@ -22,6 +22,7 @@ import (
 	"io"
 	"net/http"
 	"strings"
+	"time"
 
 	"github.com/munnerz/goautoneg"
 
@@ -116,6 +117,7 @@ func doConversionV1(convertRequest *v1.ConversionRequest, convert convertFunc) *
 }
 
 func serve(w http.ResponseWriter, r *http.Request, convert convertFunc) {
+	time.Sleep(10 * time.Millisecond)
 	var body []byte
 	if r.Body != nil {
 		if data, err := io.ReadAll(r.Body); err == nil {
diff --git a/test/integration/storageversionmigrator/storageversionmigrator_test.go b/test/integration/storageversionmigrator/storageversionmigrator_test.go
index 581064e4af0..d36b05ad8ae 100644
--- a/test/integration/storageversionmigrator/storageversionmigrator_test.go
+++ b/test/integration/storageversionmigrator/storageversionmigrator_test.go
@@ -28,6 +28,7 @@ import (
 
 	svmv1alpha1 "k8s.io/api/storagemigration/v1alpha1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	genericfeatures "k8s.io/apiserver/pkg/features"
 	encryptionconfigcontroller "k8s.io/apiserver/pkg/server/options/encryptionconfig/controller"
 	etcd3watcher "k8s.io/apiserver/pkg/storage/etcd3"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
@@ -152,6 +153,7 @@ func TestStorageVersionMigration(t *testing.T) {
 // 10. Verify RV and Generations of CRs
 // 11. Verify the list of CRs at v2 works
 func TestStorageVersionMigrationWithCRD(t *testing.T) {
+	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.ConsistentListFromCache, true)
 	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageVersionMigrator, true)
 	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, featuregate.Feature(clientgofeaturegate.InformerResourceVersion), true)
 	// decode errors are expected when using conversation webhooks
```

The test passed via the fallback case:

```
E0730 19:43:19.772016 2962489 cacher.go:896] >>>>>>>>>> Saw fallback for /stable.example.com/testcrds
[name:"apiserver_watch_cache_consistent_read_total"  help:"[ALPHA] Counter for consistent reads from cache."  type:COUNTER  metric:{label:{name:"resource"  value:"/clusterrolebindings"}  label:{name:"status"  value:"success"}  counter:{value:1  created_timestamp:{seconds:1722368549  nanos:405545523}}}  metric:{label:{name:"resource"  value:"/clusterroles"}  label:{name:"status"  value:"success"}  counter:{value:1  created_timestamp:{seconds:1722368549  nanos:400899064}}}  metric:{label:{name:"resource"  value:"/resourcequotas"}  label:{name:"status"  value:"success"}  counter:{value:5  created_timestamp:{seconds:1722368548  nanos:618619775}}}  metric:{label:{name:"resource"  value:"/services/specs"}  label:{name:"status"  value:"success"}  counter:{value:2  created_timestamp:{seconds:1722368548  nanos:419014726}}}  metric:{label:{name:"resource"  value:"/stable.example.com/testcrds"}  label:{name:"status"  value:"fallback"}  counter:{value:1  created_timestamp:{seconds:1722368599  nanos:771942121}}}]
```

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Comment by deads2k** [comment]
> I ran `make test-integration WHAT=./test/integration/storageversionmigrator KUBE_TEST_ARGS="-v -run CRD"` on [3193062](https://github.com/kubernetes/kubernetes/commit/319306218eaf275ad89c25539de0386aac616e9e) with the following diff:

Bug fix worked.

/milestone v1.31

**Comment by wojtek-t** [comment]
For posterity, this LGTM too.

**Pull request merged**